### PR TITLE
Example notebook for smdebug with EstimatorHook for TF 2.x

### PR DESCRIPTION
**Pull request by vandanavk**
*Issue #, if available:*

*Description of changes:*
Adding an example for using SageMaker Debugger with a TF2 training script that uses Keras model to Estimator API

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by vandanavk** [comment]
@Vikas-kum for review

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/src/tf_keras_to_estimator.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    optimizer = tf.keras.optimizers.Adam()
+
+    ##### Enabling SageMaker Debugger ###########
+    # wrap the optimizer so the hook can identify the gradients
+    loss = tf.losses.SparseCategoricalCrossentropy(from_logits=True)
+    model.compile(loss=loss, optimizer=optimizer, metrics=["accuracy"])
+
+    ##### Enabling SageMaker Debugger ###########
+    # creating hook
+    hook = smd.EstimatorHook(
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/src/tf_keras_to_estimator.py:68]: no 0 code change?

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/src/tf_keras_to_estimator.py:68]: there is. I can add a script in https://github.com/awslabs/amazon-sagemaker-examples/tree/master/sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/src

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/src/tf_keras_to_estimator.py:68]: done

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb [hunk 2]
```diff
@@ -0,0 +1,407 @@
+{
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>do we need this ?</p><p>By default it should pick /opt/ml/output/tensors</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: The training script has out_dir as mandatory argument.

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p><a href="https://aws.amazon.com/sagemaker/" target="_blank">Amazon SageMaker</a> is a machine learning platform to build, train, and host state-of-the-art ML and AI models. <a href="https://github.com/awslabs/sagemaker-debugger" target="_blank">Amazon SageMaker Debugger</a> offers the capability to debug machine learning models during training and identifies problems with the models in real-time.</p><p> </p><p>Experimental support for TF 2.x was initially introduced in v0.7.1 of the Debugger's smdebug library. A full description of support is available at <a href="https://github.com/awslabs/sagemaker-debugger/tree/master/docs/tensorflow.md" target="_blank">Amazon SageMaker Debugger with TensorFlow</a>.</p><p> </p><p>With the smdebug v0.9.0 release, its support has been extended to cover TF 2.x <code><a href="https://www.tensorflow.org/api_docs/python/tf/keras/estimator/model_to_estimator" target="_blank">model_to_estimator</a></code> and Estimator APIs.</p><p> </p><p>In this notebook, you will learn how to run your training job with a python script with the TF 2.x <code><a href="https://www.tensorflow.org/api_docs/python/tf/keras/estimator/model_to_estimator" target="_blank">model_to_estimator</a></code> API, debug the training job using the Debugger built-in rules,</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>We will train a TensorFlow Keras model in this notebook with Amazon Sagemaker Debugger, and monitor the training jobs with the Debugger built-in rules. The training job will be run on a pre-built <a href="https://docs.aws.amazon.com/deep-learning-containers/latest/devguide/deep-learning-containers-images.html" target="_blank">AWS Deep Learning Container</a> with Tensorflow 2.1.0 and smdebug 0.9.0 installed. </p><p> </p><p>Execute the following code cell for a one-time smdebug setup to get your notebook kernel ready for a full experience of using the Debugger features. This smdebug library provides tools to perform interactive analysis throughout the notebook.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>above the code cell, add:</p><p> </p><p>Import the AWS boto3 Python SDK, the SageMaker Python SDK, the SageMaker TensorFlow class, and other python utility libraries.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>Import the SageMaker Debugger classes for configuring hooks and rules.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>Since our demo training script <code>tf_keras_to_estimator.py</code> uses the <code>tensorflow-datasets</code> package that is not available in the pre-built container, we need to install it as a requirement for the training job. To add the installation step, we simply add the package in a <code>requirement.txt</code> file. You can add a list of required libraries to run your own training script in the same way. </p><p> </p><p>The <code>launcher.sh</code> script is provided along with this notebook, which will install the <code>tensorflow-datasets</code>package and initiate the training script. For more information about installing third-party libraries, see <a href="https://sagemaker.readthedocs.io/en/stable/frameworks/tensorflow/using_tf.html#use-third-party-libraries" target="_blank">Use third-party libraries with SageMaker Python SDK</a>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <h3>Setting up the SageMaker TensorFlow Estimator </h3><p>Now it's time to setup the SageMaker TensorFlow estimator. We will add the following Debugger-specific parameters to the estimator to enable for debugging the demo training script.</p><ul><li><strong>debugger_hook_config</strong>: This new parameter accepts a local path where you wish your tensors to be written to and also accepts the S3 bucket to store tensors. Debugger will take care of uploading these tensors transparently during execution.</li><li><strong>rules</strong>: This rules parameter will accept a list of rules you want to evaluate the tensors output and training behaviors while the training job is running. There are two types of Debugger rules:</li><li class="ql-indent-1"><strong>Built-in Rules</strong>: These are rules specially curated by the Amazon SageMaker Debugger which you can opt to evaluate your training job.</li><li class="ql-indent-1"><strong>Custom Rules</strong>: You can write your own rules as a Python source file and use them to evaluate your training job. To provide Amazon SageMaker Debugger to evaluate this rule, you have to provide the S3 location of the rule source and the evaluator image. For more information about how to create and use Debugger custom rules, see <a href="https://docs.aws.amazon.com/sagemaker/latest/dg/debugger-custom-rules.html" target="_blank">Create Debugger Custom Rules for Training Job Analysis</a>.</li></ul><h3><br></h3><h3>Using Amazon SageMaker Debugger Rules </h3><p>In this notebook, we will demonstrate how to use SageMaker built-in rules to evaluate the training script. You can find the list of SageMaker rules and the configurations best suited for using them <a href="https://github.com/awslabs/sagemaker-debugger-rulesconfig" target="_blank">here</a>. </p><p>The rules we are using in this notebook are <strong>VanishingGradient</strong> and <strong>LossNotDecreasing</strong>. As the names suggest, the rules will attempt to evaluate if there are vanishing gradients in the tensors captured by the debugging hook during training and also if the loss is not decreasing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>Once the rule job starts, this will return RuleEvaluationJabArn values. We can see the logs for the rule job in CloudWatch. To do that, we will use the following utlity functions to get the CloudWatch link to the rule job logs.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>Now that we have trained a job and looked at automated Debugger analysis through its rules. Now let's have a look at another aspect of Amazon SageMaker Debugger. It allows us to perform interactive exploration of the tensors saved in real time or after the job has finished. Here we focus on after-the-fact analysis of the above job. We import the <code>smdebug</code> library to use its tools to create <em>trials</em> that picks up the most recent training job. The following code cells show you how to fetch the path to the Debugger artifacts of the latest training job and access the saved tensors by collection names.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (mchoi8739) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p><s>We can list all the tensors that were recorded to know what we want to plot.</s> The smdebug trial class provides tools to parse the saved tensor based on names which are auto-assigned by TensorFlow. In other frameworks, the tensor names may vary and we have to use appropriate tensor names or regex based on the names of tensors such as weight, bias, gradient, input or output.</p><p>For simple examples of fetching the tensors in the following code cells, we print the total number of tensors saved for losses, weights, and gradients.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1298/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_to_estimator/tf2-keras-model-to-estimator.ipynb:1]: <p>done in latest</p>

**Review by mchoi8739 — DISMISSED** [review]

**Comment by zhaoqizqwang** [comment]
Closing this PR since it was created a long time ago and didn't pass all the checks. Feel free to open a new pull request instead.
